services/src/pro/users/oidc.rs

```diff
@@ -0,0 +1,919 @@
+        let mut user_db = self.users.write().await;
+        let pending_request = user_db.remove(&auth_code_response.state).ok_or(IllegalOIDCLoginRequest)?;
+
+        let token_response = client.exchange_code(AuthorizationCode::new(auth_code_response.code))
+            .set_pkce_verifier(pending_request.code_verifier)
+            .request_async(async_http_client)
+            .await
+            .map_err(|_token_error| OIDCLoginFailed {
+                reason: "Failed code to token exchange".to_string()
+            })?;
+
+        let id_token = token_response
+            .id_token()
+            .ok_or_else(|| OIDCLoginFailed {
+                reason: "No id token returned".to_string()
+            })?;
+
+        let claims = id_token.claims(&client.id_token_verifier(), &pending_request.nonce)
+            .map_err(|_claims_error| OIDCLoginFailed {
+                reason: "Failed to verify claims".to_string()
+            })?;
+
+        if let Some(expected_access_token_hash) = claims.access_token_hash() {
+            let actual_access_token_hash = AccessTokenHash::from_token(
+                token_response.access_token(),
+                &id_token.signing_alg()
+                    .map_err(|_signing_error| OIDCLoginFailed { reason: "Unsupported Signing Algorithm".to_string()})?
+            ).map_err(|_signing_error| OIDCLoginFailed {
+                reason: "Unsupported Signing Algorithm".to_string()
+            })?;
+            if actual_access_token_hash != *expected_access_token_hash {
+                return Err(OIDCLoginFailed {
+                    reason: "Illegal access token".to_string()
+                });
+            }
+        }
+
+        //TODO: Failing at missing claims seems unnecessary for external users, but UI expects email and real name.
+        let user = ExternalUserClaims {
+            external_id: claims.subject().clone(),
+            email: match claims.email() {
+                None => {Err(OIDCLoginFailed { reason: "Missing e-mail claim".to_string()})}
+                Some(x) => {Ok(x.to_string())}
+            }?,
+            real_name: match claims.name() {
+                None => {Err(OIDCLoginFailed { reason: "Missing name claim".to_string()})}
+                Some(x) => {Ok(x.get(None).unwrap().to_string())} //TODO: There is no Local logic.
+            }?
+        };
+
+        let validity = match token_response.expires_in(){
+            None => {Err(OIDCLoginFailed { reason: "No Duration in token".to_string() })}
+            Some(x) => {Ok(x)}
+        }?;
+
+        Ok((user, Duration::milliseconds(validity.as_millis() as i64))) //TODO: Is that cast ok, how does Geo Engine handle conversions between durations?
+    }
+
+    //TODO: Is there a better way to do that for testing? Seems hacky.
+    pub fn from_oidc_with_static_tokens(value: Oidc) -> Self {
```

Comment:
I tried to limit visibility and put all sorts of things in the test crate here: 477a5228523df29c3dd2ee0bb48d78cfa4337742

I'm getting unused methods warnings when compiling now. Probably need some help to discuss the best way to do this sort of thing in Rust.

Reply:
I guess I could write #cfg[test] over all sorts of methods/structs or maybe wrap them in another module within tests. Not sure what is best here.

Reply:
If you need it from multiple places, why not go with the `cfg`. Otherwise, you can define it within the test module as a plain function :shrug: .

Reply:
6060e382c7f6789a313bb25c734949c0e1879c86
Since i wrapped all mocking functionality in a module (mock_oidc in services/src/pro/util/tests.rs), I guess this could be moved somewhere else. Other than that, I hope this is ok now.